services/pull/merge.go

```diff
@@ -378,22 +379,55 @@ func getDiffTree(repoPath, baseBranch, headBranch string) (string, error) {
 	getDiffTreeFromBranch := func(repoPath, baseBranch, headBranch string) (string, error) {
 		var outbuf, errbuf strings.Builder
 		// Compute the diff-tree for sparse-checkout
-		if err := git.NewCommand("diff-tree", "--no-commit-id", "--name-only", "-r", "--root", baseBranch, headBranch, "--").RunInDirPipeline(repoPath, &outbuf, &errbuf); err != nil {
+		if err := git.NewCommand("diff-tree", "--no-commit-id", "--name-only", "-r", "-z", "--root", baseBranch, headBranch, "--").RunInDirPipeline(repoPath, &outbuf, &errbuf); err != nil {
 			return "", fmt.Errorf("git diff-tree [%s base:%s head:%s]: %s", repoPath, baseBranch, headBranch, errbuf.String())
 		}
 		return outbuf.String(), nil
 	}
 
+	scanNullTerminatedStrings := func(data []byte, atEOF bool) (advance int, token []byte, err error) {
+		if atEOF && len(data) == 0 {
+			return 0, nil, nil
+		}
+		if i := bytes.IndexByte(data, '\x00'); i >= 0 {
+			return i + 1, data[0:i], nil
+		}
+		if atEOF {
+			return len(data), data, nil
+		}
+		return 0, nil, nil
+	}
+
 	list, err := getDiffTreeFromBranch(repoPath, baseBranch, headBranch)
 	if err != nil {
 		return "", err
 	}
 
 	// Prefixing '/' for each entry, otherwise all files with the same name in subdirectories would be matched.
+	rightTrailingSpacesRE := regexp.MustCompile(`\ +$`)
+	optionalNPrefixRE := regexp.MustCompile(`(?:^|/)(!.+?)$`)
+
 	out := bytes.Buffer{}
 	scanner := bufio.NewScanner(strings.NewReader(list))
+	scanner.Split(scanNullTerminatedStrings)
 	for scanner.Scan() {
-		fmt.Fprintf(&out, "/%s\n", scanner.Text())
+		filepath := scanner.Text()
+
+		// Trailing spaces
+		filepath = rightTrailingSpacesRE.ReplaceAllStringFunc(filepath, func(s string) string {
+			return strings.Repeat(`\ `, len(s))
+		})
+
+		// An optional prefix !
+		filepath = optionalNPrefixRE.ReplaceAllString(filepath, `\$1`)
+
+		// * ? [
+		filepath = strings.Replace(filepath, "*", `\*`, -1)
```

Comment:
You could use [regexp.ReplaceAllString()](https://play.golang.org/p/Iyetag494tH) for this as well. However, I'm not sure about the list of special characters: why only these three specifically (I know they are used in the glob algorithm)? You'd also need to escape at least `\t` and `\\`, and I'd escape all spaces (not only trailing).

Does git enforce UTF-8 validity of the file names? (e.g. can a file with a 0x08 exist in the repo? What about ANSI encodings?)

Reply:
> why only these three specifically

git `sparse_checkout` uses the same syntax as `.gitignore` files ([link](https://git-scm.com/docs/git-read-tree#_sparse_checkout)).
`.gitignore` [pattern format](https://git-scm.com/docs/gitignore#_pattern_format).

I added `\\` to escaping symbols. `\t` works fine without escaping.